interfaces/apparmor/spec.go

```diff
@@ -371,7 +392,7 @@ func (spec *Specification) UpdateNS() []string {
 }
 
 func snippetFromLayout(layout *snap.Layout) string {
-	mountPoint := layout.Snap.ExpandSnapVariables(layout.Path)
+	mountPoint := layout.Snap.ExpandSnapMountVariables(layout.Path)
```

Comment:
We discussed how this API should evolve on IRC but to just capture it here quickly. There should be one function (ExpandSnapVariables) but it should take a mandatory flags argument that defines the _perspective_ from which the path is observed. There are two perspectives: host and snap (or outer and inner, names are TBD). For classically confined snaps the perspective is irrelevant. In addition the perspective can be fused with instance-awareness. There are very few specific places that need to use instance aware paths (that is /var/snap/name_key rather than /var/snap/name) and those must use the instance key argument. Using one function rather than two forces everyone to make a decision and figure out what kind of path perspective they are interested in.

Interestingly the inner perspective is almost always ignorant of instance awareness (only three places need to be) but the outside perspective is (perhaps always) instance aware. This can make using this call simpler as we can have a default that is sane.
```
ExpandSnapVariables(layout.Path, PerspectiveHost | InstanceUnaware) // instance-unaware path from the host 
ExpandSnapVariables(layout.Path, PerspectiveSnap | InstanceAware) // instance-aware path from the snap
ExpandSnapVariables(layout.Path, PerspectiveHost) // instance-aware path from the host 
ExpandSnapVariables(layout.Path, PerspectiveSnap) // instance-unaware path from the snap
```

Reply:
Not counting tests, those are the places where `ExpandSnapVariables()` is used:
```
snap/validate.go
417:                    sourcePath := info.ExpandSnapVariables(layout.Bind)
427:                    sourcePath := info.ExpandSnapVariables(layout.BindFile)
746:    path := layout.Snap.ExpandSnapVariables(layout.Path)
773:    mountPoint = si.ExpandSnapVariables(mountPoint)
813:            mountSource = si.ExpandSnapVariables(mountSource)
820:            if !strings.HasPrefix(mountSource, si.ExpandSnapVariables("$SNAP")) &&
821:                    !strings.HasPrefix(mountSource, si.ExpandSnapVariables("$SNAP_DATA")) &&
822:                    !strings.HasPrefix(mountSource, si.ExpandSnapVariables("$SNAP_COMMON")) {
840:            oldname = si.ExpandSnapVariables(oldname)
847:            if !strings.HasPrefix(oldname, si.ExpandSnapVariables("$SNAP")) &&
848:                    !strings.HasPrefix(oldname, si.ExpandSnapVariables("$SNAP_DATA")) &&
849:                    !strings.HasPrefix(oldname, si.ExpandSnapVariables("$SNAP_COMMON")) {

snap/info.go
424:// ExpandSnapVariables resolves $SNAP, $SNAP_DATA and $SNAP_COMMON inside the
426:func (s *Info) ExpandSnapVariables(path string) string {

interfaces/builtin/content.go
178:            return snapInfo.ExpandSnapVariables(path)
181:    return filepath.Join(snapInfo.ExpandSnapVariables("$SNAP"), path)

interfaces/apparmor/spec.go
131:            path := si.ExpandSnapVariables(l.Path)
134:                    bind := si.ExpandSnapVariables(l.Bind)
142:                    bindFile := si.ExpandSnapVariables(l.BindFile)
395:    mountPoint := layout.Snap.ExpandSnapVariables(layout.Path)

interfaces/mount/spec.go
67:     mountPoint := layout.Snap.ExpandSnapVariables(layout.Path)
72:             mountSource := layout.Snap.ExpandSnapVariables(layout.Bind)
77:             mountSource := layout.Snap.ExpandSnapVariables(layout.BindFile)
88:             oldname := layout.Snap.ExpandSnapVariables(layout.Symlink)
```

The current use cases are in the context of layouts and content interface. All assume the inside view of the snap's mount namespace. Maybe we could go YAGNI about it, leave the function as it is now and reevaluate any changes when those are actually needed. WDYT?

Reply:
Agreed, if there's no need let's just leave it alone for now. We have enough trouble to solve to get parallel installs off the ground, and we can easily revisit this when it becomes an actual issue.